File: phases/17-testing-and-verification/08-contracts-and-design-by-contract/code/main.py
#!/usr/bin/env python3
from __future__ import annotations


def transfer(src_balance: int, dst_balance: int, amount: int) -> tuple[int, int]:
    if amount <= 0:
        raise ValueError("precondition: amount > 0")
    if src_balance < amount:
        raise ValueError("precondition: source has enough funds")

    new_src = src_balance - amount
    new_dst = dst_balance + amount

    if new_src < 0:
        raise AssertionError("invariant violated: src balance non-negative")
    if new_src + new_dst != src_balance + dst_balance:
        raise AssertionError("postcondition violated: conservation")

    return new_src, new_dst


def main() -> None:
    print(transfer(100, 20, 30))


if __name__ == "__main__":
    main()

File: phases/17-testing-and-verification/08-contracts-and-design-by-contract/code/main.rs
#[derive(Debug)]
enum TransferError {
    NonPositiveAmount,
    InsufficientFunds,
    InvariantViolation,
}

fn transfer(src: i64, dst: i64, amount: i64) -> Result<(i64, i64), TransferError> {
    if amount <= 0 {
        return Err(TransferError::NonPositiveAmount);
    }
    if src < amount {
        return Err(TransferError::InsufficientFunds);
    }

    let new_src = src - amount;
    let new_dst = dst + amount;

    if new_src < 0 || new_src + new_dst != src + dst {
        return Err(TransferError::InvariantViolation);
    }
    Ok((new_src, new_dst))
}

fn main() {
    println!("{:?}", transfer(100, 20, 30));
}
